src/gui/tray/Window.qml

```diff
@@ -624,56 +618,184 @@ Window {
                     }
-                    ToolTip.visible: hovered
-                    ToolTip.delay: 1000
-                    ToolTip.text: qsTr("Open share dialog")
-                    onClicked: Systray.openShareDialog(displayPath,absolutePath)
 
-                    Accessible.role: Accessible.Button
-                    Accessible.name: qsTr("Share %1").arg(displayPath)
-                    Accessible.onPressAction: shareButton.clicked()
-                }
+                    Button {
+                        id: actionPrimaryButton
 
-                Button {
-                    id: replyButton
-                    anchors.right: activityItem.right
+                        readonly property int actionIndex: 0
 
-                    Layout.preferredWidth: (objectType == "chat" || objectType == "call") ? parent.height : 0
-                    Layout.preferredHeight: parent.height
-                    Layout.alignment: Qt.AlignRight
-                    flat: true
-                    hoverEnabled: true
-                    visible: (objectType == "chat" || objectType == "call") ? true : false
-                    display: AbstractButton.IconOnly
-                    icon.source: "qrc:///client/theme/reply.svg"
-                    icon.color: "transparent"
-                    background: Rectangle {
-                        color: parent.hovered ? Style.lightHover : "transparent"
+                        Layout.preferredWidth: visible ? parent.height : 0
+                        Layout.preferredHeight: parent.height
+                        flat: true
+                        hoverEnabled: true
+                        visible: links.length > 0 && activityItem.totalActions < 3
+                        display: AbstractButton.IconOnly
+                        icon.source: activityActionsLayout.actionButtonIcon(actionIndex)
+                        icon.color: "transparent"
+                        background: Rectangle {
+                            color: parent.hovered ? Style.lightHover : "transparent"
+                        }
+                        ToolTip.visible: hovered
+                        ToolTip.delay: 1000
+                        ToolTip.text: links[actionIndex]._label
+                        onClicked: activityModel.handleAction(model.index, actionIndex)
+
+                        Accessible.role: Accessible.Button
+                        Accessible.name: links[actionIndex]._label
+                        Accessible.onPressAction: actionPrimaryButton.clicked()
                     }
-                    ToolTip.visible: hovered
-                    ToolTip.delay: 1000
-                    ToolTip.text: qsTr("Open Talk")
-                    onClicked: Qt.openUrlExternally(link)
 
-                    Accessible.role: Accessible.Button
-                    Accessible.name: qsTr("Open Talk %1").arg(link)
-                    Accessible.onPressAction: replyButton.clicked()
+                    Button {

```

Comment:
Those buttons feel like they could be somehow created from a Repeater... if you find an obvious way to do this right now go for it. Otherwise let's keep it for a later refactoring once we start splitting proper components out of this giant file.

Reply:
@er-vin Realized that I've altered most of Button's properties, and also I've needed very different behavior for displaying icon/text in the same button. Moved it to a separate file and created a button from scratch with MouseArea.